cmd/aws-s3-csi-controller/csicontroller/reconciler.go

```diff
@@ -0,0 +1,338 @@
+		// and they might got deleted once we try to re-process them again.
+		if apierrors.IsNotFound(err) {
+			log.Info("Pod not found - ignoring")
+			return reconcile.Result{}, nil
+		}
+		log.Error(err, "Failed to get Pod")
+		return reconcile.Result{}, err
+	}
+
+	if r.isMountpointPod(pod) {
+		return r.reconcileMountpointPod(ctx, pod)
+	}
+
+	return r.reconcileWorkloadPod(ctx, pod)
+}
+
+// reconcileMountpointPod reconciles given Mountpoint `pod`, and deletes it if its completed.
+func (r *Reconciler) reconcileMountpointPod(ctx context.Context, pod *corev1.Pod) (reconcile.Result, error) {
+	log := logf.FromContext(ctx).WithValues("mountpointPod", pod.Name)
+
+	switch pod.Status.Phase {
+	case corev1.PodPending:
+		log.V(debugLevel).Info("Pod pending to be scheduled")
+	case corev1.PodRunning:
+		log.V(debugLevel).Info("Pod is running")
+	case corev1.PodSucceeded:
+		err := r.deleteMountpointPod(ctx, pod)
+		if err != nil {
+			log.Error(err, "Failed to delete succeeded Pod")
+			return reconcile.Result{}, err
+		}
+		log.Info("Pod succeeded and successfully deleted")
+	case corev1.PodFailed:
+		// TODO: We should probably delete failed Pods after some time to trigger a retry on the whole operation.
+		// 		 Maybe just returning a `reconcile.Result{RequeueAfter: ...}`
+		// 	     and deleting in next cycle would be a good way?
+		log.Info("Pod failed", "reason", pod.Status.Reason)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// reconcileWorkloadPod reconciles given workload `pod` to spawn a Mountpoint Pod to provide a volume for it if needed.
+func (r *Reconciler) reconcileWorkloadPod(ctx context.Context, pod *corev1.Pod) (reconcile.Result, error) {
+	log := logf.FromContext(ctx).WithValues("pod", types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name})
+
+	if pod.Spec.NodeName == "" {
+		log.V(debugLevel).Info("Pod is not scheduled to a node yet - ignoring")
+		return reconcile.Result{}, nil
+	}
+
+	if len(pod.Spec.Volumes) == 0 {
+		log.V(debugLevel).Info("Pod has no volumes - ignoring")
+		return reconcile.Result{}, nil
+	}
+
+	var requeue bool
+	var errs []error
+
+	for _, vol := range pod.Spec.Volumes {
+		podPVC := vol.PersistentVolumeClaim
+		if podPVC == nil {
+			continue
+		}
+
+		// If PVC has no bound PVs yet, `getBoundPVForPodClaim` will return `errPVCIsNotBoundToAPV`.
+		// In this case we'll just return `reconcile.Result{Requeue: true}` here, which will bubble up to the
+		// original `Reconcile` call and will cause a retry for this Pod with an exponential backoff.
+		pvc, pv, err := r.getBoundPVForPodClaim(ctx, pod, podPVC)
+		if err != nil {
+			if errors.Is(err, errPVCIsNotBoundToAPV) {
+				requeue = true
+			} else {
+				errs = append(errs, err)
+			}
+			continue
+		}
+
+		csiSpec := extractCSISpecFromPV(pv)
+		if csiSpec == nil {
+			continue
+		}
+
+		log.V(debugLevel).Info("Found bound PV for PVC", "pvc", pvc.Name, "volumeName", pv.Name)
+
+		err = r.spawnOrDeleteMountpointPodIfNeeded(ctx, pod, pvc, pv, csiSpec)
```

Comment:
This can be called multiple times for the same volume. I don't think this is a problem though, but is it worth migrating this outside the for loop and only calling if we have no requeue and no errors? That way we could call in parallel for each PVC, but that might just be premature optimisation.

Reply:
I think I'd like this to be called even some other volume for this Pod requires requeueing, so one poisonous volume won't block other volumes to progress. Yeah, I think calling this in parallel would be a pre-mature optimization and would also be problematic if we want to share Mountpoint instances between multiple volumes. They would be racing to spawn Mountpoint Pods and that might cause duplicate Pods.